File: 3486.longest-special-path-ii.py
import sys
import heapq

# Increase recursion depth to handle deep trees
sys.setrecursionlimit(100000)

class Solution:
    def longestSpecialPath(self, edges: List[List[int]], nums: List[int]) -> List[int]:
        n = len(nums)
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))
            
        # pos[val] stores the list of depths where val appears in the current path
        pos = {} 
        
        # Max-heaps for lazy removal to maintain the set S of "penultimate" indices
        # S stores the indices that trigger a duplicate (i.e., the index of the second-to-last occurrence)
        # We use negative values to simulate a max-heap with Python's heapq
        max_heap = [] 
        remove_heap = [] 
        
        self.max_len = -1
        self.min_nodes = 0
        path_dists = [] 
        
        def get_max_s():
            while remove_heap and max_heap[0] == remove_heap[0]:
                heapq.heappop(max_heap)
                heapq.heappop(remove_heap)
            return -max_heap[0] if max_heap else -1
            
        def get_second_max_s():
            first = get_max_s()
            if first == -1: return -1
            
            # Temporarily remove the max to find the second max
            heapq.heappop(max_heap)
            second = get_max_s()
            # Restore the max
            heapq.heappush(max_heap, -first)
            return second

        def add_to_s(val):
            heapq.heappush(max_heap, -val)
            
        def remove_from_s(val):
            heapq.heappush(remove_heap, -val)

        def dfs(u, p, depth, current_dist, current_path_start_limit):
            path_dists.append(current_dist)
            val = nums[u]
            if val not in pos: pos[val] = []
            pos[val].append(depth)
            
            # Constraint 1: No value appears >= 3 times
            # If a value appears for the 3rd time (or more), the path must start after the 3rd-to-last occurrence
            new_path_start_limit = current_path_start_limit
            if len(pos[val]) >= 3:
                new_path_start_limit = max(new_path_start_limit, pos[val][-3] + 1)
                
            # Manage S for Constraint 2: At most one value appears 2 times
            added_trigger = -1
            removed_trigger = -1
            
            if len(pos[val]) >= 2:
                # The penultimate occurrence is now a trigger for a duplicate
                added_trigger = pos[val][-2]
                add_to_s(added_trigger)
                
                # If there were already >= 2 occurrences before this one, the OLD penultimate
                # (which is now antepenultimate) was in S. We must remove it because it's no longer the trigger.
                # The trigger is always the second-to-last occurrence.
                if len(pos[val]) > 2:
                    removed_trigger = pos[val][-3]
                    remove_from_s(removed_trigger)
                    
            s_second_max = get_second_max_s()
            # Start index must be > second largest trigger point to allow at most 1 duplicate
            valid_start = max(new_path_start_limit, s_second_max + 1)
            
            path_len = current_dist - path_dists[valid_start]
            num_nodes = depth - valid_start + 1
            
            if path_len > self.max_len:
                self.max_len = path_len
                self.min_nodes = num_nodes
            elif path_len == self.max_len:
                self.min_nodes = min(self.min_nodes, num_nodes)
                
            for v, w in adj[u]:
                if v != p:
                    dfs(v, u, depth + 1, current_dist + w, new_path_start_limit)
            
            # Backtrack state changes
            if added_trigger != -1: remove_from_s(added_trigger)
            if removed_trigger != -1: add_to_s(removed_trigger)
            pos[val].pop()
            path_dists.pop()

        dfs(0, -1, 0, 0, 0)
        return [self.max_len, self.min_nodes]

File: 3486.longest-special-path-ii.cpp
#include <bits/stdc++.h>
using namespace std;

// @lc app=leetcode id=3486 lang=cpp
//
// [3486] Longest Special Path II
//

// @lc code=start
class Solution {
public:
    vector<int> longestSpecialPath(vector<vector<int>>& edges, vector<int>& nums) {
        int n = (int)nums.size();
        vector<vector<pair<int,int>>> g(n);
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            g[u].push_back({v, w});
            g[v].push_back({u, w});
        }

        int maxVal = 0;
        for (int x : nums) maxVal = max(maxVal, x);
        vector<vector<int>> occ(maxVal + 1);

        multiset<int> ms2; // second-from-last depth for values with >=2 occurrences
        multiset<int> ms3; // third-from-last depth for values with >=3 occurrences

        vector<int> nodeStack;
        vector<long long> distStack;
        nodeStack.reserve(n);
        distStack.reserve(n);

        long long bestLen = 0;
        int bestNodes = 1;

        auto addValueAtDepth = [&](int val, int depth) {
            auto &vec = occ[val];
            int k = (int)vec.size();

            // Remove old contributions
            if (k >= 2) {
                auto it = ms2.find(vec[k - 2]);
                if (it != ms2.end()) ms2.erase(it);
            }
            if (k >= 3) {
                auto it = ms3.find(vec[k - 3]);
                if (it != ms3.end()) ms3.erase(it);
            }

            // Insert new contributions after push (based on previous elements)
            if (k >= 1) {
                ms2.insert(vec[k - 1]);
            }
            if (k >= 2) {
                ms3.insert(vec[k - 2]);
            }

            vec.push_back(depth);
        };

        auto removeValueAtDepth = [&](int val) {
            auto &vec = occ[val];
            int k = (int)vec.size();
            // k >= 1

            // Remove current contributions (before pop)
            if (k >= 2) {
                auto it = ms2.find(vec[k - 2]);
                if (it != ms2.end()) ms2.erase(it);
            }
            if (k >= 3) {
                auto it = ms3.find(vec[k - 3]);
                if (it != ms3.end()) ms3.erase(it);
            }

            // Insert new contributions after pop
            if (k >= 3) {
                ms2.insert(vec[k - 3]);
            }
            if (k >= 4) {
                ms3.insert(vec[k - 4]);
            }

            vec.pop_back();
        };

        auto currentLeftBoundary = [&]() -> int {
            int l = 0;
            if (!ms3.empty()) {
                l = max(l, *ms3.rbegin() + 1);
            }
            if ((int)ms2.size() >= 2) {
                auto rit = ms2.rbegin();
                ++rit; // second largest
                l = max(l, *rit + 1);
            }
            return l;
        };

        struct Frame {
            int u, p;
            int it;
            long long dist;
        };

        vector<Frame> st;
        st.reserve(n);

        // enter root
        st.push_back({0, -1, 0, 0LL});
        {
            int depth = 0;
            nodeStack.push_back(0);
            distStack.push_back(0LL);
            addValueAtDepth(nums[0], depth);

            int l = currentLeftBoundary();
            long long len = distStack[depth] - distStack[l];
            int nodes = depth - l + 1;
            if (len > bestLen) {
                bestLen = len;
                bestNodes = nodes;
            } else if (len == bestLen) {
                bestNodes = min(bestNodes, nodes);
            }
        }

        while (!st.empty()) {
            Frame &fr = st.back();
            int u = fr.u;
            if (fr.it == (int)g[u].size()) {
                // exit node
                int depth = (int)nodeStack.size() - 1;
                removeValueAtDepth(nums[u]);
                nodeStack.pop_back();
                distStack.pop_back();
                st.pop_back();
                continue;
            }

            auto [v, w] = g[u][fr.it++];
            if (v == fr.p) continue;

            long long ndist = fr.dist + w;
            st.push_back({v, u, 0, ndist});

            int depth = (int)nodeStack.size();
            nodeStack.push_back(v);
            distStack.push_back(ndist);
            addValueAtDepth(nums[v], depth);

            int l = currentLeftBoundary();
            long long len = distStack[depth] - distStack[l];
            int nodes = depth - l + 1;
            if (len > bestLen) {
                bestLen = len;
                bestNodes = nodes;
            } else if (len == bestLen) {
                bestNodes = min(bestNodes, nodes);
            }
        }

        return { (int)bestLen, bestNodes };
    }
};
// @lc code=end
